model/task/task.go

```diff
@@ -1439,30 +1412,59 @@ func ResetTasks(taskIds []string) error {
 		bson.M{
 			IdKey: bson.M{"$in": taskIds},
 		},
-		bson.M{
-			"$set": bson.M{
-				ActivatedKey:           true,
-				SecretKey:              utility.RandomString(),
-				HostIdKey:              "",
-				StatusKey:              evergreen.TaskUndispatched,
-				DispatchTimeKey:        utility.ZeroTime,
-				StartTimeKey:           utility.ZeroTime,
-				ScheduledTimeKey:       utility.ZeroTime,
-				FinishTimeKey:          utility.ZeroTime,
-				DependenciesMetTimeKey: utility.ZeroTime,
-				TimeTakenKey:           utility.ZeroTime,
-			},
-			"$unset": bson.M{
-				DetailsKey:           "",
-				HasCedarResultsKey:   "",
-				ResetWhenFinishedKey: "",
-			},
-		},
+		resetTaskUpdate(nil),
 	)
 
 	return err
 }
 
+func resetTaskUpdate(t *Task) bson.M {
+	newSecret := utility.RandomString()
+	now := time.Now()
+	if t != nil {
+		t.Activated = true
+		t.ActivatedTime = now
+		t.Secret = newSecret
+		t.HostId = ""
+		t.Status = evergreen.TaskUndispatched
+		t.DispatchTime = utility.ZeroTime
+		t.StartTime = utility.ZeroTime
+		t.ScheduledTime = utility.ZeroTime
+		t.FinishTime = utility.ZeroTime
+		t.DependenciesMetTime = utility.ZeroTime
+		t.TimeTaken = 0
+		t.LastHeartbeat = utility.ZeroTime
+		t.Details = apimodels.TaskEndDetail{}
+		t.HasCedarResults = false
+		t.ResetWhenFinished = false
+		t.HostId = ""
+		t.AgentVersion = ""
+	}
+	update := bson.M{
+		"$set": bson.M{
+			ActivatedKey:           true,
+			ActivatedTimeKey:       now,
+			SecretKey:              newSecret,
+			HostIdKey:              "",
```

Comment:
Why $set and not $unset?

Reply:
$unset is on the omitempty fields, $set is on the ones that always have a value